### ISSUE TITLE
Spurious pipeline state changes with dynamic-tee-vsink.c

### ISSUE THREAD

**Issue by SylvainGarrigues** [sdroege/gst-snippets#1]
Thanks so much for this example and your work in general.

Quick question here: I just added some state change debug to the pipeline bus loop:
```
--- dynamic-tee-vsink.c.old	2023-08-20 21:34:20
+++ dynamic-tee-vsink.c	2023-08-20 21:36:36
@@ -45,6 +45,15 @@
 message_cb (GstBus * bus, GstMessage * message, gpointer user_data)
 {
   switch (GST_MESSAGE_TYPE (message)) {
+    case GST_MESSAGE_STATE_CHANGED:{
+      if (GST_MESSAGE_SRC (message) == GST_OBJECT (pipeline)) {
+        GstState old_state, new_state, pending_state;
+        gst_message_parse_state_changed (message, &old_state, &new_state, &pending_state);
+        g_print ("Pipeline state changed from %s to %s (pending: %s)",
+          gst_element_state_get_name (old_state), gst_element_state_get_name (new_state), gst_element_state_get_name (pending_state));
+      }
+      break;
+    }
     case GST_MESSAGE_ERROR:{
       GError *err = NULL;
       gchar *name, *debug = NULL;
``` 

I am surprised to see pipeline state changes while the playback doesn't seem to be paused:
```
% ./dynamic-tee-vsink somevideo.mp4
Pipeline state changed from NULL to READY (pending: PLAYING)
Pipeline state changed from READY to PAUSED (pending: PLAYING)
Pipeline state changed from PAUSED to PLAYING (pending: VOID_PENDING)
add
added
Pipeline state changed from PAUSED to PAUSED (pending: PAUSED)
Pipeline state changed from PAUSED to PAUSED (pending: PLAYING)
Pipeline state changed from PAUSED to PLAYING (pending: VOID_PENDING)
remove
removed
add
added
Pipeline state changed from PAUSED to PAUSED (pending: PAUSED)
Pipeline state changed from PAUSED to PAUSED (pending: PLAYING)
Pipeline state changed from PAUSED to PLAYING (pending: VOID_PENDING)
remove
removed
```

Is there any reason the pipeline gets its state changed during the add operations (see the 3 lines  after `added`), and why are there PAUSED to PAUSED state changes?

**Comment by SylvainGarrigues** [comment]
(by the way, I obviously get the same issue with the equivalent Rust program)

**Comment by sdroege** [comment]
This happens because sinks with `async=true` are added to the pipeline (a non-live pipeline). This temporarily causes the pipeline to lose its `PLAYING` state until the new sink is pre-rolled and `PLAYING`.

Similarly, setting a sink from `PLAYING` to `PAUSED` makes it wait until it is pre-rolled again and has the same effect on the pipeline state (the sink would go to `PAUSED` pending `PAUSED`).

This behaviour can be disabled by setting `async=false` on the sinks.

**Comment by SylvainGarrigues** [comment]
Thank you very much for taking the time to answer. Still, I don’t see any `PLAYING -> PAUSED` transition for the pipeline before the `PAUSED -> PAUSED` one, looks like no such `GST_MESSAGE_STATE_CHANGED` state change message is posted to the bus: is that part of the `causes the pipeline *to [literally] lose* its PLAYING state’? That is what surprises me most.

**Comment by sdroege** [comment]
Yes, it directly jumps back without going `PLAYING -> PAUSED` first. It's a inconsistent part of the state change design and something to be fixed in the next GStreamer version (2.0) together with various other design fixes for state changes (and other things).
